### PR TITLE
[Model] Fused mm preprocess normalisation on the Device

### PR DESCRIPTION
**Pull request by noooop**
## Purpose

This is one of my favorite CV tricks.

1. When normalize is computed on the GPU, it can be fused into subsequent operators, or even if left unfused, it takes almost no time.
2. Currently, image data transmission from the entrypoint to the engine core and then to the GPU uses uint8, rather than the model dtype bf16, and is only converted to bf16 on the GPU before entering input normal.

Moreover, this can be applied to all vision models.

### Speedup mm preprocess 

https://github.com/noooop/snippet/blob/main/benchmarks/mm_device_do_normalize/test.py

```
disable mm_device_do_normalize:
Qwen2VLImageProcessor preprocess
[<PIL.Image.Image image mode=RGB size=1024x1024 at 0x7D422A3F77D0>]
{'return_tensors': 'pt'}
preprocess 12.21515999895928 ms
e2e: 0.1540803940006299

enable mm_device_do_normalize:
Qwen2VLImageProcessor preprocess
[<PIL.Image.Image image mode=RGB size=1024x1024 at 0x73DF97EDEF00>]
{'do_rescale': False, 'return_tensors': 'pt', 'do_normalize': False}
preprocess 6.270558998949127 ms
e2e: 0.14389813100024185
```

Qwen2VLImageProcessor speed on 1024x1024 image has nearly doubled. 12ms -> 6ms

### vllm bench

```
Mean TTFT (ms):
with mm_device_do_normalize: 3459.52
without mm_device_do_normalize: 3809.70
VLLM_USE_RUST_FRONTEND: 3078.26

Median TTFT (ms):
with mm_device_do_normalize: 3057.37
without mm_device_do_normalize: 3612.82
VLLM_USE_RUST_FRONTEND: 2021.02
```

Absolute strength beats any strategy. 一力降十会

This optimization should be orthogonal to Rush, so the Rush frontend can also add support for it to further improve performance.

<details> 

```
vllm bench serve \
--backend openai-chat \
--base-url http://localhost:8000 \
--endpoint /v1/chat/completions \
--model Qwen/Qwen2-VL-2B-Instruct \
--dataset-name random-mm \
--num-prompts 100 \
--random-output-len 1 


vllm serve Qwen/Qwen2-VL-2B-Instruct

============ Serving Benchmark Result ============
Successful requests:                     100
Failed requests:                         0
Benchmark duration (s):                  8.15
Total input tokens:                      168350
Total generated tokens:                  100
Request throughput (req/s):              12.27
Output token throughput (tok/s):         12.27
Peak output token throughput (tok/s):    50.00
Peak concurrent requests:                100.00
Total token throughput (tok/s):          20675.97
---------------Time to First Token----------------
Mean TTFT (ms):                          3459.52
Median TTFT (ms):                        3057.37
P99 TTFT (ms):                           7993.54
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          0.00
Median TPOT (ms):                        0.00
P99 TPOT (ms):                           0.00
---------------Inter-token Latency----------------
Mean ITL (ms):                           0.01
Median ITL (ms):                         0.01
P99 ITL (ms):                            0.03
==================================================


vllm serve Qwen/Qwen2-VL-2B-Instruct --no-mm-device-do-normalize

============ Serving Benchmark Result ============
Successful requests:                     100
Failed requests:                         0
Benchmark duration (s):                  8.15
Total input tokens:                      168350
Total generated tokens:                  100
Request throughput (req/s):              12.28
Output token throughput (tok/s):         12.28
Peak output token throughput (tok/s):    44.00
Peak concurrent requests:                100.00
Total token throughput (tok/s):          20677.63
---------------Time to First Token----------------
Mean TTFT (ms):                          3809.70
Median TTFT (ms):                        3612.82
P99 TTFT (ms):                           7991.42
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          0.00
Median TPOT (ms):                        0.00
P99 TPOT (ms):                           0.00
---------------Inter-token Latency----------------
Mean ITL (ms):                           0.02
Median ITL (ms):                         0.02
P99 ITL (ms):                            0.06
==================================================

VLLM_USE_RUST_FRONTEND=1 vllm serve Qwen/Qwen2-VL-2B-Instruct

============ Serving Benchmark Result ============
Successful requests:                     100
Failed requests:                         0
Benchmark duration (s):                  8.28
Total input tokens:                      168350
Total generated tokens:                  100
Request throughput (req/s):              12.07
Output token throughput (tok/s):         12.07
Peak output token throughput (tok/s):    78.00
Peak concurrent requests:                100.00
Total token throughput (tok/s):          20336.37
---------------Time to First Token----------------
Mean TTFT (ms):                          3078.26
Median TTFT (ms):                        2021.02
P99 TTFT (ms):                           8154.79
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          0.00
Median TPOT (ms):                        0.00
P99 TPOT (ms):                           0.00
---------------Inter-token Latency----------------
Mean ITL (ms):                           0.01
Median ITL (ms):                         0.01
P99 ITL (ms):                            0.03
==================================================
```
</details>

### PPL test

The numerical results are exactly the same.

```
Model: Qwen/Qwen2-VL-2B-Instruct
mm_device_do_normalize: True
VLLM: dtype:torch.bfloat16 head_dtype:torch.bfloat16 41096328.0
Transformers: Constant 41081356.0
Difference (%): 0.03644475610785584

Model: Qwen/Qwen2-VL-2B-Instruct
mm_device_do_normalize: False
VLLM: dtype:torch.bfloat16 head_dtype:torch.bfloat16 41096328.0
Transformers: Constant 41081356.0
Difference (%): 0.03644475610785584


Model: Qwen/Qwen2.5-VL-3B-Instruct
mm_device_do_normalize: True
VLLM: dtype:torch.bfloat16 head_dtype:torch.bfloat16 18406568.0
Transformers: Constant 18330016.0
Difference (%): 0.4176319322361748 

Model: Qwen/Qwen2.5-VL-3B-Instruct
mm_device_do_normalize: False
VLLM: dtype:torch.bfloat16 head_dtype:torch.bfloat16 18406568.0
Transformers: Constant 18330016.0
Difference (%): 0.4176319322361748
```

## Test Plan

tests/models/multimodal/processing/test_qwen2_vl.py::test_mm_device_do_normalize
tests/models/multimodal/generation_ppl_test/test_qwen.py

## Test Result
pass

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

File: vllm/model_executor/models/qwen2_5_vl.py [hunk 1]
```diff
@@ -1375,6 +1377,11 @@ def __init__(self, *, vllm_config: VllmConfig, prefix: str = ""):
             self.language_model.make_empty_intermediate_tensors
         )
 
+        if multimodal_config.mm_device_do_normalize:
+            self.input_norm = make_input_norm(Qwen2VLImageProcessor)
+        else:
+            self.input_norm = None
```
Comment (Isotr0py) [inline vllm/model_executor/models/qwen2_5_vl.py]:
```suggestion
        self.input_norm = make_input_norm(Qwen2VLImageProcessor) if multimodal_config.mm_device_do_normalize else nn.Identity()
```
I think we can use `nn.Identity()`. And let's move this under `self._mark_tower_model` context?

**Comment by DarkLight1337** [comment]
How much does this actually improve E2E though? You moved the computation to model execution, of course the image processor is going to take less time

**Review by Isotr0py — COMMENTED** [review]
I think it's better to provide e2e TTFT improvement comparing rust frontend as well. Because we also make a normalize+to_tensor fusion there.

**Comment by noooop** [comment]
@DarkLight1337 @Isotr0py 

```
Mean TTFT (ms):
with mm_device_do_normalize: 3459.52
without mm_device_do_normalize: 3809.70
VLLM_USE_RUST_FRONTEND: 3078.26

Median TTFT (ms):
with mm_device_do_normalize: 3057.37
without mm_device_do_normalize: 3612.82
VLLM_USE_RUST_FRONTEND: 2021.02
```

Absolute strength beats any strategy. 一力降十会

This optimization should be orthogonal to Rush, so the Rush frontend can also add support for it to further improve performance.

File: vllm/config/model.py [hunk 2]
```diff
@@ -755,6 +757,8 @@ def __post_init__(
                 video_pruning_method=video_pruning_method,
                 mm_tensor_ipc=mm_tensor_ipc,
                 mm_ipc_gpu_memory_gb=mm_ipc_gpu_memory_gb,
+                mm_device_do_normalize=mm_device_do_normalize
+                and self._model_info.supports_mm_device_do_normalize,
```
Comment (DarkLight1337) [inline vllm/config/model.py]: I think we should add logs to indicate when `mm_device_do_normalize` gets disabled forcefully

File: vllm/transformers_utils/processor.py [hunk 3]
```diff
@@ -582,3 +584,43 @@ def call_hf_processor_mm_only(
         data={**audio_inputs, **images_inputs, **videos_inputs},
         tensor_type=kwargs.get("return_tensors"),
     )
+
+
+def make_input_norm(processor) -> nn.BatchNorm1d:
```
Comment (DarkLight1337) [inline vllm/transformers_utils/processor.py]: I don't think this is the proper way to do it, you're just reading the default values from the processor class, instead of the actual HF processor instance that's being loaded from the config

**Comment by DarkLight1337** [comment]
Btw, I think part of the reason this helps is the fact that we batch together the processed MM inputs in the engine before model's forward pass, not just because the calculation is done on GPU

**Comment by DarkLight1337** [comment]
I have been meaning to look into batched MM processing for a while, but haven't had the time to do it, and things recently got more complicated with the addition of Rust frontend...

File: vllm/transformers_utils/processor.py [hunk 4]
```diff
@@ -582,3 +584,52 @@ def call_hf_processor_mm_only(
         data={**audio_inputs, **images_inputs, **videos_inputs},
         tensor_type=kwargs.get("return_tensors"),
     )
+
+
+def make_input_norm(processor: type[BaseImageProcessor]) -> nn.BatchNorm1d:
+    image_mean = getattr(processor, "image_mean", None)
+    image_std = getattr(processor, "image_std", None)
+    rescale_factor = getattr(processor, "rescale_factor", None)
```
Comment (Isotr0py) [inline vllm/transformers_utils/processor.py]: Some processor can be initialized from custom remote code, which is difficult to be imported. Perhaps we can read the values from [preprocessor_config.json](https://huggingface.co/Qwen/Qwen3.6-35B-A3B/blob/main/preprocessor_config.json#L9-L18)

File: vllm/transformers_utils/processor.py [hunk 5]
```diff
@@ -582,3 +584,52 @@ def call_hf_processor_mm_only(
...
+
+    return bn
+
+
+def maybe_do_input_norm(
+    grid_thw: torch.Tensor,
+    input_norm: nn.Module | None,
+    visual_dtype: torch.dtype,
+    channel: int = 3,
+) -> torch.Tensor:
```
Comment (Isotr0py) [inline vllm/transformers_utils/processor.py]: Let's move these two functions to `vllm/model_executor/models/vision.py`

**Review by Isotr0py — APPROVED** [review]
LGTM now

File: vllm/config/model.py [hunk 6]
```diff
@@ -736,6 +738,25 @@ def __post_init__(
                 )
                 mm_encoder_tp_mode = "weights"
 
+            if envs.VLLM_USE_RUST_FRONTEND:
```
Comment (DarkLight1337) [inline vllm/config/model.py]: Let's add a warning if the user enables this for a configuration that's not supported, so that the user doesn't get confused about why the setting isn't working

**Review by DarkLight1337 — COMMENTED** [review]
And we should update the optimization guide in the docs as well

File: vllm/model_executor/models/vision.py [hunk 7]
```diff
@@ -580,3 +582,55 @@ def run_dp_sharded_mrope_vision_model(
...
     )
     return out_embeddings
+
+
+def make_input_norm(model_config: "ModelConfig") -> nn.BatchNorm1d:
+    model = model_config.model
+    revision = model_config.revision
+    config = get_video_processor_config(model, revision=revision)
+
+    image_mean = config.get("image_mean", None)
```
Comment (DarkLight1337) [inline vllm/model_executor/models/vision.py]: A bit of an edge case, but what if the processor class has a default `do_rescale=False` or `do_normalize=False`, or if they are set in HF config files or in the original `mm_processor_kwargs`?

Comment (noooop) [reply vllm/model_executor/models/vision.py]: I will manually test the other models and add supports_mm_device_do_normalize to the supported architectures.

Comment (noooop) [reply vllm/model_executor/models/vision.py]: Can't guarantee all same‑architecture HF models share the same config – edge cases possible.

┓( ´∀` )┏

Comment (DarkLight1337) [reply vllm/model_executor/models/vision.py]: I think we could just import the class and check the default value, this shouldn't need to be model specific

Comment (noooop) [reply vllm/model_executor/models/vision.py]: I think it’s pretty unlikely that preprocessor_config.json would contain image_mean and image_std if do_rescale or do_normalize is False.

Comment (noooop) [reply vllm/model_executor/models/vision.py]: Added assert do_rescale and do_normalize

Comment (DarkLight1337) [reply vllm/model_executor/models/vision.py]: Also why do we need to assert this? Can't we just return `nn.Identity` or default to a rescale/normalize factor of 1 if no normalize/rescale should be done?

File: vllm/model_executor/models/vision.py [hunk 8]
```diff
@@ -580,3 +582,68 @@ def run_dp_sharded_mrope_vision_model(
...
+
+    do_rescale = config.get("do_rescale", True)
+    do_normalize = config.get("do_normalize", True)
+
+    if not do_rescale and not do_normalize:
+        return nn.Identity()
+
+    image_mean = config.get("image_mean", None)
+    image_std = config.get("image_std", None)
+    rescale_factor = config.get("rescale_factor", 1 / 255)
```
Comment (DarkLight1337) [inline vllm/model_executor/models/vision.py]: I think it's better to get the proper defaults from the transfomers class. Is that not possible to get from the config?

Comment (noooop) [reply vllm/model_executor/models/vision.py]: https://github.com/huggingface/transformers/blob/main/src/transformers/utils/constants.py

```
IMAGENET_DEFAULT_MEAN = [0.485, 0.456, 0.406]
IMAGENET_DEFAULT_STD = [0.229, 0.224, 0.225]
IMAGENET_STANDARD_MEAN = [0.5, 0.5, 0.5]
IMAGENET_STANDARD_STD = [0.5, 0.5, 0.5]
OPENAI_CLIP_MEAN = [0.48145466, 0.4578275, 0.40821073]
OPENAI_CLIP_STD = [0.26862954, 0.26130258, 0.27577711]
```

There are many commonly used MEAN and STD values, so we’d better fail fast to avoid using an inappropriate mean and std.

Comment (noooop) [reply vllm/model_executor/models/vision.py]: qwen2-vl using OPENAI_CLIP but Qwen3.5 using IMAGENET_STANDARD

https://huggingface.co/Qwen/Qwen3.5-4B/blob/main/preprocessor_config.json
https://huggingface.co/Qwen/Qwen2-VL-2B-Instruct/blob/main/preprocessor_config.json

Comment (DarkLight1337) [reply vllm/model_executor/models/vision.py]: I mean we could try to get the `AutoProcessor` class from the config, so we could import it and get the default values from the class definition

File: vllm/model_executor/models/vision.py [hunk 9]
```diff
@@ -580,3 +582,82 @@ def run_dp_sharded_mrope_vision_model(
...
+    model = model_config.model
+    revision = model_config.revision
+    config = get_processor_config(model, revision=revision)
+
+    do_rescale = config.get("do_rescale", None)
+    do_normalize = config.get("do_normalize", None)
+
+    image_mean = config.get("image_mean", None)
+    image_std = config.get("image_std", None)
+    rescale_factor = config.get("rescale_factor", 1 / 255)
```
Comment (DarkLight1337) [inline vllm/model_executor/models/vision.py]: Default of `rescale_factor` should also be None

**Review by DarkLight1337 — APPROVED** [review]
LGTM now, thanks

File: vllm/model_executor/models/vision.py [hunk 10]
```diff
@@ -580,3 +583,134 @@ def run_dp_sharded_mrope_vision_model(
...
+        image_mean: list[float],
+        image_std: list[float],
+        rescale_factor: float,
+        channel: int = 3,
+    ):
+        super().__init__()
+
+        self.channel = channel
+
+        image_mean_tensor = torch.tensor(image_mean, dtype=torch.float32) * (
```
Comment (DarkLight1337) [inline vllm/model_executor/models/vision.py]: Use a dtype argument to init, to better resemble pytorch's own batch norm

File: vllm/model_executor/models/vision.py [hunk 11]
```diff
@@ -580,3 +583,136 @@ def run_dp_sharded_mrope_vision_model(
...
+        if self.is_identity:
+            return grid_thw.to(visual_dtype)
+
+        assert grid_thw.ndim == 2
+        patches, size = grid_thw.shape
+        patch_size = size // self.channel
+
+        grid_thw = grid_thw.view(patches, self.channel, patch_size)
+        grid_thw = F.batch_norm(
+            grid_thw.to(torch.float32),
```
Comment (DarkLight1337) [inline vllm/model_executor/models/vision.py]:
```suggestion
            grid_thw.to(self.dtype),
```

Comment (DarkLight1337) [reply vllm/model_executor/models/vision.py]: And remember to set the attribute

Comment (noooop) [reply vllm/model_executor/models/vision.py]: updating, The internet is a bit slow today.

**Review by DarkLight1337 — APPROVED** [review]
